### ISSUE TITLE
backup: relative paths handled incorrectly

### ISSUE THREAD

**Issue by chastings667** [restic/restic#3131]
Output of `restic version`
--------------------------
restic 0.11.0 compiled with go1.15.3 on linux/amd64

How did you run restic exactly?
-------------------------------

+ /home/USER/bin/restic-0.11.0 init --repo ./repo
+ /home/USER/bin/restic-0.11.0 backup --repo ./repo ./data -v -v
+ /home/USER/bin/restic-0.11.0 backup --repo ./repo ././data -v -v
+ /home/USER/bin/restic-0.11.0 backup --repo ./repo ../testing/./data -v -v
+ /home/USER/bin/restic-0.11.0 backup --repo ./repo ../testing/../testing/./data -v -v
+ /home/USER/bin/restic-0.11.0 backup --repo ./repo /home/USER/src/restic/testing/./data -v -v

Please see attached logfile for full cmds and output. But the above is enough to recreate the problem, on an existing repo.

[log-sanitized.txt](https://github.com/restic/restic/files/5610299/log-sanitized.txt)

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

Directory on local filesystem.

Expected behavior
-----------------

All of these backup paths should create snapshots with the same file contents. The restic ls of files should show the same files.

[an aside: is ls supposed to show the absolute paths to files contained in a snapshot?]

The /data/ dir tree has two files in it:

a/a0.bin
a/a1.bin

Actual behavior
---------------

The five snapshots show three different file contents:

1,2 ==>
/data
/data/a
/data/a/a0.bin
/data/a/a1.bin

3,4 ==>
/testing
/testing/data
/testing/data/a
/testing/data/a/a0.bin
/testing/data/a/a1.bin

5 ==>
/home
/home/USER
/home/USER/src
/home/USER/src/restic
/home/USER/src/restic/testing
/home/USER/src/restic/testing/data
/home/USER/src/restic/testing/data/a
/home/USER/src/restic/testing/data/a/a0.bin
/home/USER/src/restic/testing/data/a/a1.bin

Logfile above will have more details.

Steps to reproduce the behavior
-------------------------------

Set res= to your binary.
Change USER and HOST, and a few paths to something sane.

[test-sanitized.sh.txt](https://github.com/restic/restic/files/5610308/test-sanitized.sh.txt)

This script should repro the issue.

Or just follow along with the bash logfile above, the lines starting with + are cmds.

Do you have any idea what may have caused this?
-----------------------------------------------
There's something wonky with the relative path handling. 

Do you have an idea how to solve the issue?
-------------------------------------------
Don't use relative paths? 😃 

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
I haven't suffered any data loss so not directly.
But it's nice to have a direct communication line to the people who can fix issues like this one!
And... it's great to have useful free software out there.

**Comment by fd0** [comment]
Hey, this is an interesting issue. First, what's your use case? What are you trying to do?

Restic used to work like you seem to expect it to work, but that was **much** worse.

> [an aside: is ls supposed to show the absolute paths to files contained in a snapshot?]

No. Please be aware that `restic snapshots` displays the paths/files as passed to restic, after resolving them to absolute paths.

Ok, if I understood you correctly, you'd expect the following commands:

```
/home/USER/bin/restic-0.11.0 backup --repo ./repo ./data -v -v
/home/USER/bin/restic-0.11.0 backup --repo ./repo ../testing/data -v -v
```

to produce the same structure in the snapshot:

```
/data
/data/a
/data/a/a0.bin
/data/a/a1.bin
```

Is that correct?

I left out the `./` cases because the path gets normalized before backup even starts, so `./data` is treated the same as `././data`.

I rewrote the archiver code some time ago and decided to implement behavior which mimics what `tar` does:

```
$ tar cvf /tmp/data1.tar ./data
./data/
./data/file2
./data/file1

$ tar cvf /tmp/data2.tar ../testing/data
tar: Removing leading `../' from member names
../testing/data/
../testing/data/file2
../testing/data/file1
```

As you can see it'll also create different structures in the `tar` file. In restic, snapshots reference a tree, containing either more trees or files (or other stuff such as symlinks). There's no concept of a directory referencing something in a tree above the current one...

**Comment by fd0** [comment]
I'm unable to find the issue, but one of the corner cases was the following: Let's say you have the following structure:
```
.
├── user1
│   └── data
│       ├── file
│       └── otherfile
└── user2
    └── data
        └── file3
```

And you run `cd user1` and run `restic backup ./data ../user2/data`. How should the structure look like in the snapshot?

**Comment by chastings667** [comment]
> Hey, this is an interesting issue. First, what's your use case? What are you trying to do?

There isn't a particular use case. I was testing something unrelated (multiple parent snapshots) and noticed that for these relative paths, sometimes parent selection didn't go as expected. I looked into it and noticed that file paths (shown with ls) were not what I expected.

> Restic used to work similar to how you seem to expect it to work, but that was much worse.

You mean storing absolute paths in the tree? I'm curious, what problems did this cause?

> Ok, if I understood you correctly, you'd expect the following commands:
> to produce the same structure in the snapshot:

I'm not exactly sure WHAT I expect it to produce, but I expected it to be the same since both relative paths evaluate to the same directory.

> Honest question: Why do you think restic should work like that?

This is kind of difficult to answer. I guess I see backup software as data-focused. You point it at some data, and it stores it. No matter how you got to that data (relative paths, symlinks, whatever) it's the same data, and should be stored and presented in the same way. So I was surprised that the five backup paths above, which point to the same data, end up with three different representations in snapshots. I'm having trouble understanding why that is ever the right thing to do.

> I rewrote the archiver code some time ago and decided to implement behavior which mimics what tar does, so restic loosely > follows the Principle of Least Astonishment:

I hadn't heard of this specifically but I completely agree with the idea. Remember though that you have at least two different perspectives: the developers, who are probably less frequently astonished by the software they write and maintain, and new users (like myself) who might come in with (perhaps inapplicable) preconceptions.

That said, maybe I just haven't used tar enough to expect its semantics elsewhere.  It's good that there's an existing model for how this should work, even if it's not what I expected.

**Comment by chastings667** [comment]
> And you run cd user1 and run restic backup ./data ../user2/data. How should the structure look like in the snapshot?

Well, these rel paths point to different data so there could be any number of representations. I don't see why the snapshot paths need to match what the user specified with rel paths. I don't see the paths specified as "my files better start with this string" and more of "look here for the stuff I want to back up".

So, absolute paths would work:

/parent/user1/data/file
/parent/user1/data/otherfile
/parent/user2/data/file3

(where /parent/ starts at root, and is the full path to user1 and user2)

Or I you could traverse upwards to find a common node in the dir hierarchy?

user1/data/file
user1/data/otherfile
user2/data/file3

How would the current algorithm handle this case?

And, using your example, how would the following be stored?

`cd user1 && restic backup ./data ../user1/data ../user1/../user1/data ../user2/data ../user1/../user2/data`

Common-parent or absolute handle this sensibly, but I really don't know what would happen with the current algorithm,

Thanks for the discussion, this is interesting!

**Comment by fd0** [comment]
> You mean storing absolute paths in the tree? I'm curious, what problems did this cause?

No, in the first version of the archiver (used before 0.9.0 AFAIR) just used the last path element as the top-level element for the snapshot, so `restic backup /home/user/foo` led to a snapshot that had `/foo` as the top-level item. That led to all sorts of problems (see #549) and was especially surprising for users.

> So I was surprised that the five backup paths above, which point to the same data, end up with three different representations in snapshots. I'm having trouble understanding why that is ever the right thing to do.

That's interesting. I wanted to give users a bit of control as to how the structure looks like in the snapshot. Users requested that very often, sometimes there's a path prefix that they'd like to not be present in the snapshot (like `/mnt/zfs-snapshot`) or something else. It's also a use case question: when you're saving a complete system (`restic backup /`) it will preserve the absolute paths, if you just want to save a directory within your home, then you can decide if you want to keep the full path (`restic backup /home/user/data`) or if you want to have `data` be the top-level item (`cd $HOME; restic backup data`).

When designing restic, I took a lot of inspiration from existing tools, like `tar`, `git`, and sometimes `rsync`. The handling of absolute versus relative paths is very similar to how `tar` handles them, so I'm very familiar with it, and I suspect many other users, too.

> Remember though that you have at least two different perspectives: the developers, who are probably less frequently astonished by the software they write and maintain, and new users (like myself) who might come in with (perhaps inapplicable) preconceptions.

Oh yes, absolutely! I oftentimes need to remind myself that it's not people coming to us and demand that something needs to be done differently (although some issues/posts sometimes come across that way), but that's people using the software with a completely different experiences and assumptions. Decoupling the questions people have from the perception of being critized is hard work :wink:

Most of the time there's no single "correct" way to do something, and changing anything within restic will inevitably lead to other people being astonished that its behavior has changed :smile:

> So, absolute paths would work:

And you can have that, just pass absolute paths to restic, e.g. `restic backup $PWD/data $PWD/../user2/data`. It's just a matter of figuring out how to represent relative paths within a snapshot. 

> Or I you could traverse upwards to find a common node in the dir hierarchy?

Yep, I thought about that when I rewrote the archiver code, and decided against it. There's no "right" way here, so I chose to do what `tar` does (or something very similar to it). Most people who have worked with systems or administration before are familiar with it, so I thought it'd be good to behave similarly.

> And, using your example, how would the following be stored?
>
> `cd user1 && restic backup ./data ../user1/data ../user1/../user1/data ../user2/data ../user1/../user2/data`
>
> Common-parent or absolute handle this sensibly, but I really don't know what would happen with the current algorithm

I had to look it up:
 1. Start with the set of paths: `./data ../user1/data ../user1/../user1/data ../user2/data ../user1/../user2/data`
 2. Resolve relative paths semantically if possible: `./data ../user1/data ../user1/data ../user2/data ../user2/data`
 3. Remove duplicates: `./data ../user1/data ../user2/data`
 4. Create a snapshot:
      * top-level `data` dir in the snapshot contains `./data`
      * `user1/data` dir in the snapshot contains `./data` again, just at a different path (it's deduplicated so there's only a tiny amount of overhead)
      * `user2/data` dir in the snapshot contains `../user2/data`

> Thanks for the discussion, this is interesting!

Likewise!

**Comment by chastings667** [comment]
Yes, I think "unexpectedly" is a better description. And now my google-fu yields some better results:

https://forum.restic.net/t/backing-up-restoring-relative-paths/744/2

Is the tar-like behavior mentioned in the docs anywhere? I'd guess it would be in the backup section:

https://restic.readthedocs.io/en/stable/040_backup.html#backing-up-special-items-and-metadata

but "relative" is only mentioned in passing and "tar" isn't at all.

**Comment by dmotte** [comment]
For those who are still facing this frustrating issue in 2021... I have found a workaround. I always invoke restic backup like this:

```bash
restic backup -vr /repo "$(realpath .)"
```

or, if you use it on different systems, you can do this instead:

```bash
cd /my/folder; restic backup -vr /repo .; cd -
```

🔥 **Update 2025-07-07**: see **new solution**: https://github.com/restic/restic/issues/2092#issuecomment-2632525938
